internal/queue/queue.go

```diff
@@ -168,3 +176,45 @@ func (q *SendRecvQueues) Delete(name string, shutdown bool) error {
 	delete(q.queues, name)
 	return nil
 }
+
+// GetWithContext is a wrapper around the workqueue's Get method, and it
+// uses a condition variable to wait for an item to be available. It also
+// handles the case where the context is cancelled while waiting for an
+// item to be available.
+func GetWithContext(q workqueue.TypedRateLimitingInterface[*event.Event], ctx context.Context) (*event.Event, bool) {
+	bq, ok := q.(*boundedQueue)
+	if !ok {
+		return nil, false
+	}
+
+	done := make(chan struct{})
+	bq.notify.L.Lock()
+	defer func() {
+		bq.notify.L.Unlock()
+		close(done)
+	}()
+
+	// Goroutine to signal on context cancellation
+	go func() {
+		select {
+		case <-ctx.Done():
+			bq.notify.L.Lock()
+			bq.notify.Signal()
+			bq.notify.L.Unlock()
+		case <-done:
+			// If function exits early, prevent leak
+		}
+	}()
+
+	// Suspend until an item is available or context is cancelled
+	for q.Len() == 0 && ctx.Err() == nil {
+		bq.notify.Wait()
+	}
```

Comment:
I think using a `for with select` to check if a context is done is more simple. I'm using a condition variable so that we don't spin continuously in a for loop and poll if the queue has new items.

```
for {
 select {
    case <-ctx.Done:
         // return from function
    default:
       if q.Len() == 0:
            continue
       time.Sleep(100 * millisecond)
  }
}
```

I don't think polling would be a big issue now. So, I'm fine with replacing the condition variable with the simple for select logic. Or even better, if there is another easy way to handle this case.

Reply:
Have you considered using a simple channel instead of a condition to synchronize, as pointed out in the [docs](https://pkg.go.dev/sync#Cond)?

Reply:
I was trying to extend the [condition variable logic](https://github.com/kubernetes/client-go/blob/master/util/workqueue/queue.go#L277) used by workqueue's Get() to be aware of ctx. But I agree that channels are more idiomatic. We could use a buffered channel (non-blocking) to signal when an item is added to the queue.